File: python/src/fem.py
import numpy as np
from concurrent import futures

class Fem():
    def __init__(self,dof,nodes,elements,faults,materials):
        self.nnode = len(nodes)
        self.nelem = len(elements)
        self.dof = dof

        self.nodes = nodes
        self.elements = elements
        self.faults = faults
        self.materials = materials

        self.input_elements = []
        self.free_nodes = []
        self.fixed_nodes = []
        self.connected_elements = []
        self.fault_elements = []

    # ======================================================================= #
    def set_init(self):
        self._set_mesh()
        self._set_initial_condition()
        self._set_initial_matrix()

    # ---------------------------------------
    def _set_mesh(self):
        for element in self.elements:
            nodes = []
            for inode in element.inode:
                if self.nodes[inode].id == inode:
                    nodes += [self.nodes[inode]]
                else:
                    for n in self.nodes:
                        if n.id == inode:
                            nodes += [n]
                            break
            element.set_nodes(nodes)

            if element.material_id < 0:
                element.set_material(None)
            else:
                if self.materials[element.material_id].id == element.material_id:
                    material = self.materials[element.material_id]
                else:
                    for m in self.materials:
                        if m.id == element.material_id:
                            material = m
                            break
                element.set_material(material)

            if "input" in element.style:
                self.input_elements += [element]
            if "connect" in element.style:
                self.connected_elements += [element]
            if "fault" in element.style:
                self.fault_elements += [element]

    # ---------------------------------------
    def _set_initial_condition(self):
        for node in self.nodes:
            node.set_initial_condition()

            if np.count_nonzero(node.freedom) == self.dof:
                self.free_nodes += [node]
            else:
                self.fixed_nodes += [node]

        for element in self.elements:
            element.set_pointer_list()

    # ---------------------------------------
    def _set_initial_matrix(self):
        for element in self.elements:
            element.set_xn()
            element.mk_local_matrix_init(self.dof)
            element.mk_local_matrix()
            element.mk_local_vector()

            id = 0
            for node in element.nodes:
                for i in range(self.dof):
                    node.mass[i] += element.M_diag[id]
                    node.c[i] += element.C_diag[id]
                    node.k[i] += element.K_diag[id]
                    node.static_force[i] += element.force[id]
                    id += 1

    # ======================================================================= #
    def set_output(self,outputs):
        output_node_list,output_element_list = outputs

        self.output_nnode = len(output_node_list)
        self.output_nodes = [None] * self.output_nnode
        for id,inode in enumerate(output_node_list):
            self.output_nodes[id] = self.nodes[inode]

        self.output_nelem = len(output_element_list)
        self.output_elements = [None] * self.output_nelem
        for id,ielem in enumerate(output_element_list):
            self.output_elements[id] = self.elements[ielem]

    # ======================================================================= #
    def set_initial_fault(self):
        self.fault_p_elements = []
        self.fault_m_elements = []

        for fault in self.faults:
            fault.set_initial_condition0(self.elements)
            self.fault_p_elements += [fault.pelement]
            self.fault_m_elements += [fault.melement]

        for fault in self.faults:
            fault.set_initial_condition1(self.elements)


    # ======================================================================= #
    def update_init(self,dt):
        for node in self.nodes:
            node.inv_mc = 1.0 / (node.mass[:] + 0.5*dt*node.c[:])
            node.mass_inv_mc = node.mass[:]*node.inv_mc[:]
            node.c_inv_mc = node.c[:]*node.inv_mc[:]*0.5*dt
            node.dtdt_inv_mc = dt*dt*node.inv_mc[:]

        self.dt = dt
        self.inv_dt2 = 1./(2.*dt)
        self.inv_dtdt = 1./(dt*dt)

    # ======================================================================= #
    def update_matrix(self):
        for node in self.node_set:
            self._update_matrix_node_init(node)
        for element in self.element_set:
            self._update_matrix_set_elements(element)

    # ---------------------------------------
    def _update_matrix_node_init(self,node):
        node.mass = np.zeros(self.dof,dtype=np.float64)
        node.c    = np.zeros(self.dof,dtype=np.float64)
        node.dynamic_force = np.zeros(self.dof,dtype=np.float64)

    def _update_matrix_set_elements(self,element):
        element.set_xn()
        element.mk_local_update()

        id = 0
        for node in element.nodes:
            for i in range(self.dof):
                node.mass[i] += element.M_diag[id]
                node.c[i] += element.C_diag[id]
                node.dynamic_force[i] += element.force[id]
                id += 1

    def _update_matrix_set_nodes(self,node):
        node.inv_mc = 1.0 / (node.mass[:] + 0.5*self.dt*node.c[:])
        node.mass_inv_mc = node.mass[:]*node.inv_mc[:]
        node.c_inv_mc = node.c[:]*node.inv_mc[:]*0.5*self.dt
        node.dtdt_inv_mc = self.dt*self.dt*node.inv_mc[:]

    # ======================================================================= #
    def update_time(self,acc0,vel0=None,input_wave=False):
        for node in self.node_set:
            node.dynamic_force = np.zeros(self.dof,dtype=np.float64)
            self._update_time_node_init(node)

        if input_wave:
            for element in self.input_element_set:
                self._update_time_input_wave(element,vel0)
        else:
            for element in self.element_set:
                self._update_bodyforce(element,acc0)

        for element in self.element_set:
            element.mk_ku_cv()

        for node in self.free_node_set:
            self._update_time_set_free_nodes(node)
        for node in self.fixed_node_set:
            self._update_time_set_fixed_nodes(node)

        for element in self.connected_element_set:
            self._update_time_set_connected_elements_(element)

        for element in self.output_element_set:
            element.calc_stress()

    # ======================================================================= #
    def update_time_slip(self,slip0):
        for node in self.node_set:
            node.dynamic_force = np.zeros(self.dof,dtype=np.float64)

        for node in self.node_set:
            self._update_time_node_init(node)

        for element in self.element_set:
            element.mk_ku_cv()

        for node in self.free_node_set:
            self._update_time_set_free_nodes(node)
        for node in self.fixed_node_set:
            self._update_time_set_fixed_nodes(node)

        for element in self.slip_joint_node_elements:
            self._update_time_set_slip_joint_node_elements_(element,slip0)

        for element in self.connected_element_set:
            self._update_time_set_connected_elements_(element)

        for element in self.output_element_set:
            element.calc_stress()

    # ======================================================================= #
    def update_time_source(self,sources,slip0):
        for node in self.nodes:
            node.dynamic_force = np.zeros(self.dof,dtype=np.float64)

        for node in self.nodes:
            self._update_time_node_init(node)

        for source in sources:
            self._update_time_source(source,slip0)

        for element in self.elements:
            element.mk_ku_cv()

        for node in self.free_nodes:
            self._update_time_set_free_nodes(node)
        for node in self.fixed_nodes:
            self._update_time_set_fixed_nodes(node)

        for element in self.connected_elements:
            self._update_time_set_connected_elements_(element)

        for element in self.output_elements:
            element.calc_stress()

    # ======================================================================= #
    def update_time_dynamic_fault(self):
        for node in self.nodes:
            node.dynamic_force = np.zeros(self.dof,dtype=np.float64)
            self._update_time_node_init(node)

        for element in self.elements:
            element.mk_ku_cv()

        for element in self.fault_m_elements:
            self._update_time_fault_m_elements(element)
        for element in self.fault_p_elements:
            self._update_time_fault_p_elements(element)

        for node in self.free_nodes:
            self._update_time_set_free_nodes(node)
        for node in self.fixed_nodes:
            self._update_time_set_fixed_nodes(node)

        for fault in self.faults:
            fault.update_friction(self.dt)

        for fault in self.faults:
            fault.calc_traction(self.elements)

        for fault in self.faults:
            fault.update_rupture(self.elements)

        for element in self.output_elements:
            element.calc_stress()

    # ---------------------------------------
    def _update_time_node_init(self,node):
        node.force = -node.dynamic_force.copy()

    def _update_time_input_wave(self,element,vel0):
        cv = element.C @ np.tile(vel0,element.nnode)
        for i in range(element.nnode):
            i0 = self.dof*i
            element.nodes[i].force[:] -= 2*cv[i0:i0+self.dof]

    def _update_bodyforce(self,element,acc0):
        element.mk_bodyforce(acc0)
        for i in range(element.nnode):
            i0 = self.dof*i
            element.nodes[i].force[:] -= element.force[i0:i0+self.dof]

    def _update_time_source(self,source,slip0):
        id = source.element_id
        self.elements[id].mk_source(source,slip0)
        for i in range(self.elements[id].nnode):
            i0 = self.dof*i
            self.elements[id].nodes[i].force[:] -= self.elements[id].force[i0:i0+self.dof]

    def _update_time_set_free_nodes(self,node):
        u = np.copy(node.u)
        node.u[:] = node.mass_inv_mc*(2.*u-node.um) + node.c_inv_mc*node.um - node.dtdt_inv_mc*node.force
        node.v[:] = (node.u - node.um) * self.inv_dt2
        node.a[:] = (node.u - 2.*u + node.um) * self.inv_dtdt
        node.um = u

    def _update_time_set_fixed_nodes(self,node):
        u = np.copy(node.u)
        for i in range(self.dof):
            if node.freedom[i] == 0:
                node.u[i] = 0.0
                node.v[i] = 0.0
                node.a[i] = 0.0
            else:
                node.u[i] = node.mass_inv_mc[i]*(2.*u[i]-node.um[i]) + node.c_inv_mc[i]*node.um[i] - node.dtdt_inv_mc[i]*node.force[i]
                node.v[i] = (node.u[i] - node.um[i]) * self.inv_dt2
                node.a[i] = (node.u[i] - 2.*u[i] + node.um[i]) * self.inv_dtdt
        node.um = u

    def _update_time_set_connected_elements_(self,element):
        u = np.zeros_like(element.nodes[0].u)
        a = np.zeros_like(element.nodes[0].a)
        for node in element.node_set:
            u[:] += node.u[:]
            a[:] += node.a[:]
        for node in element.node_set:
            node.u[:] = u[:]/element.nnode
            node.a[:] = a[:]/element.nnode

    def _update_time_set_slip_joint_node_elements_(self,element,slip0):
        slip = np.array([0.0,0.5*slip0])
        element.nodes[0].u[:] =  element.R.T @ slip
        element.nodes[1].u[:] = -element.R.T @ slip

    def _update_time_fault_p_elements(self,element):
        Tinput = np.array([0.0,-element.traction,0.0])
        T = element.R.T @ Tinput
        element.mk_T(T)

    def _update_time_fault_m_elements(self,element):
        Tinput = np.array([0.0,element.traction,0.0])
        T = element.R.T @ Tinput
        element.mk_T(T)

    # ======================================================================= #
    def print_all(self):
        for node in self.nodes:
            node.print()
        for element in self.elements:
            element.print()

File: cpp/src/fem.cpp
#include "all.h"
#include <Eigen/Core>
#include "node.h"
#include "material.h"
#include "element_style.h"
#include "element.h"
#include "source.h"
#include "fem.h"

using EV = Eigen::VectorXd ;

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
Fem::Fem (size_t dof, std::vector<Node> nodes,
                std::vector<Element> elements,
                std::vector<Material> materials)
  {
    this->nnode = nodes.size();
    this->nelem = elements.size();
    this->dof = dof;

    this->nodes = nodes;
    this->elements = elements;
    this->materials = materials;
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::set_init() {
    this->_set_mesh();
    this->_set_initial_condition();
    this->_set_initial_matrix();
  }

// ----------------------------- //
void Fem::_set_mesh() {
    for (auto& element : this->elements) {

      std::vector<Node*> nodes_p(element.nnode);
      for (size_t i = 0 ; i < element.nnode ; i++ ){
        size_t inode = element.inode[i];

        if (this->nodes[inode].id == inode) {
          nodes_p[i] = &this->nodes[inode];
        } else {
          for (auto& node : this->nodes) {
            if (node.id == inode) {
              nodes_p[i] = &node;
              break;
            }
          }
        }
      }

      element.set_nodes(nodes_p);

      Material* material_p = nullptr;
      if (this->materials[element.material_id].id == element.material_id) {
        material_p = &this->materials[element.material_id];
      } else {
        for (auto& material : this->materials) {
          if (material.id == element.material_id) {
            material_p = &material;
            break;
          }
        }
      }

      element.set_material(material_p);

      if (element.style.find("solid") != std::string::npos) {
        this->solid_elements_p.push_back(&element);
      }
      if (element.style.find("visco") != std::string::npos) {
        this->visco_elements_p.push_back(&element);
      }
      if (element.style.find("input") != std::string::npos) {
        this->input_elements_p.push_back(&element);
      }
      if (element.style.find("connect") != std::string::npos) {
        this->connected_elements_p.push_back(&element);
      }
    }
  }

// ----------------------------- //
void Fem::_set_initial_condition() {
    for (auto& node : this->nodes) {
      node.set_initial_condition();

      size_t total_dof = std::accumulate(node.freedom.begin(),node.freedom.end(),0);
      if (total_dof == dof) {
        this->free_nodes_p.push_back(&node);
      } else {
        this->fixed_nodes_p.push_back(&node);
      }
    }
  }

// ----------------------------- //
void Fem::_set_initial_matrix(){
    for (auto& element : this->elements) {
      element.set_xn();
      element.mk_local_matrix_init(this->dof);
      element.mk_local_matrix();
      element.mk_local_vector();

      size_t id = 0;
      for (size_t inode = 0 ; inode < element.nnode ; inode++) {
        for (size_t i = 0 ; i < this->dof ; i++) {
          element.nodes_p[inode]->mass[i] += element.M_diag[id];
          element.nodes_p[inode]->c[i] += element.C_diag[id];
          id++;
        }
      }

      // if (element.style.find("input") != std::string::npos ||
      //     element.style.find("visco") != std::string::npos) {
      //   size_t id = 0;
      //   for (size_t inode = 0 ; inode < element.nnode ; inode++) {
      //     for (size_t i = 0 ; i < this->dof ; i++) {
      //       element.nodes_p[inode]->c[i] += element.C_diag[id];
      //       id++;
      //     }
      //   }
      // }
    }
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::set_output(std::tuple<std::vector<size_t>, std::vector<size_t>> outputs) {
    auto [output_node_list, output_element_list] = outputs;

    this->output_nnode = output_node_list.size();
    for (size_t inode = 0 ; inode < this->output_nnode ; inode++) {
      size_t id = output_node_list[inode];
      this->output_nodes_p.push_back(&this->nodes[id]);
    }

    this->output_nelem = output_element_list.size();
    for (size_t ielem = 0 ; ielem < this->output_nelem ; ielem++) {
      size_t id = output_element_list[ielem];
      this->output_elements_p.push_back(&this->elements[id]);
    }
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::update_init(const double dt) {
    this->dt = dt;
    this->inv_dt2 = 1.0/(2.0*dt);
    this->inv_dtdt = 1.0/(dt*dt);
  }

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::update_time(const EV acc0, const EV vel0, const bool input_wave) {
    if(input_wave) {
      this->update_time_input(vel0);
    } else {
      exit(1);
    }
  }

// ------------------------------------------------------------------- //
void Fem::update_time_input(const EV vel0) {
    for (auto& node : this->nodes) {
      node.force = EV::Zero(this->dof);
    }

    for (auto& element_p : this->input_elements_p) {
      element_p->update_inputwave(vel0);
    }

    for (auto& element : this->elements) {
      element.mk_ku_cv();
    }

    this->_update_time_set_free_nodes();
    this->_update_time_set_fixed_nodes();
    this->_update_time_set_connected_elements();

    for (auto& element_p : this->output_elements_p) {
      element_p->calc_stress();
    }
  }

// ------------------------------------------------------------------- //
void Fem::update_time_source(const std::vector<Source> sources, const double slip0) {
    for (auto& node : this->nodes) {
      node.force = EV::Zero(this->dof);
    }

    this->_update_time_source(sources,slip0);

    // for (auto& element : this->elements) {
    //   element.mk_ku_cv();
    // }

    for (auto& element_p : this->solid_elements_p) {
      element_p->mk_ku();
    }
    for (auto& element_p : this->visco_elements_p) {
      element_p->mk_ku_cv();
    }

    this->_update_time_set_free_nodes();
    this->_update_time_set_fixed_nodes();
    this->_update_time_set_connected_elements();

    for (auto& element_p : this->output_elements_p) {
      element_p->calc_stress();
    }
  }

void Fem::_update_time_source(const std::vector<Source> sources, const double slip0) {
  for (auto& source : sources){
    size_t id = source.element_id;

    this->elements[id].mk_source(source.dn,source.strain_tensor,slip0);

    for (size_t inode = 0 ; inode < this->elements[id].nnode ; inode++){
      size_t i0 = inode*this->dof;
      for (size_t i = 0 ; i < this->dof ; i++) {
        this->elements[id].nodes_p[inode]->force(i) -= this->elements[id].force[i0+i];
      }
    }
  }
}

// ------------------------------------------------------------------- //
// ------------------------------------------------------------------- //
void Fem::_update_time_set_free_nodes() {
    for (auto& node_p : this->free_nodes_p) {
      EV u = node_p->u;
      for (size_t i = 0 ; i < node_p->dof ; i++) {
        double inv_mc = 1.0 / (node_p->mass[i] + 0.5*this->dt*node_p->c[i]);
        double mass_inv_mc = node_p->mass[i] * inv_mc;
        double c_inv_mc = node_p->c[i] * inv_mc * 0.5*this->dt;
        double dtdt_inv_mc = this->dt*this->dt*inv_mc;

        node_p->u[i] = mass_inv_mc*(2.0*u[i] - node_p->um[i])
                + c_inv_mc*node_p->um[i] - dtdt_inv_mc*node_p->force[i];

      }
      node_p->v = (node_p->u - node_p->um) * this->inv_dt2;
      node_p->um = u;
    }
  }

void Fem::_update_time_set_fixed_nodes() {
    for (auto& node_p : this->fixed_nodes_p) {
      EV u = node_p->u;
      for (size_t i = 0 ; i < node_p->dof ; i++) {
        if (node_p->freedom[i] == 0) {
          node_p->u[i] = 0.0;
        } else {
          double inv_mc = 1.0 / (node_p->mass[i] + 0.5*this->dt*node_p->c[i]);
          double mass_inv_mc = node_p->mass[i] * inv_mc;
          double c_inv_mc = node_p->c[i] * inv_mc * 0.5*this->dt;
          double dtdt_inv_mc = this->dt*this->dt*inv_mc;

          node_p->u[i] = mass_inv_mc*(2.0*u[i] - node_p->um[i])
                  + c_inv_mc*node_p->um[i] - dtdt_inv_mc*node_p->force[i];
        }
      }
      node_p->v = (node_p->u - node_p->um) * this->inv_dt2;
      node_p->um = u;
    }
  }

void Fem::_update_time_set_connected_elements() {
    for (auto& element_p : this->connected_elements_p) {
      EV u = EV::Zero(element_p->dof);
      for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
        u += element_p->nodes_p[inode]->u;
      }
      for (size_t inode = 0 ; inode < element_p->nnode ; inode++) {
        element_p->nodes_p[inode]->u = u/element_p->nnode;
      }
    }
  }
